### ISSUE TITLE
Unused locals doesn't mark interpolated strings.

### ISSUE THREAD

**Issue by brettfo** [dotnet/roslyn#1484]
### CODE:

``` C#
class C
{
    void M()
    {
        var unused1 = ""; // marked as unused
        var unused2 = $""; // not marked as unused
    }
}
```

### EXPECTED:
`unused1` and `unused2` should be marked as unused locals.

### RESULT:
Only `unused1` is marked.

**Comment by gafter** [comment]
This behaves identical to

```cs
        var unused2 = String.Format("");
```

as specified. Closing as by design.

**Comment by Przemyslaw-W** [comment]
And why unused2 is not marked as unused in the

var unused2 = String.Format("");

case?


2015-03-23 22:57 GMT+01:00 Neal Gafter <notifications@github.com>:

**Comment by gafter** [comment]
Because the user might be intending it to keep alive some memory.

**Comment by Przemyslaw-W** [comment]
But does it force this diagnostic to be disabled in such relative rare and
not standard use case? I think it would be better to report it - majority
of users would benefit. If one really needs to keep the code, he can always
suppress this particular occurence of this diagnostic

2015-03-24 3:15 GMT+01:00 Neal Gafter <notifications@github.com>:

**Comment by gafter** [comment]
We never report that diagnostic for computed reference values. I agree that it would be helpful to expand the scope of when that is reported, but doing so requires support for warning waves #1580 to avoid being a breaking change.

**Comment by ljw1004** [comment]
@gafter I disagree with your conclusions.

It does *not* require "warning waves" in order to give a warning for interpolated strings. I expect that interpolated strings should give a warning, because to a typical reader of C#, they look closer to strings than to a call to String.Format.

**Comment by AnthonyDGreen** [comment]
In lieu of a warning why node fade the variable declaration as unnecessary? That wouldn't be a break but would provide much of the same value in the way that we now report unused usings.

**Comment by gafter** [comment]
@AnthonyDGreen It isn't unnecessary as it has a side effect (of calling an API, as required by the language spec).

**Comment by gafter** [comment]
@ljw1004 The interpolated string feature is specified to behave as a call to `string.Format`. The corresponding code with `string.Format` also lacks the diagnostic. So I am closing it as by design.

If you'd like a diagnostic for both cases, please file a new issue.

**Comment by Epaminaidos** [comment]
I cannot believe that you are serious about this. Detecting unused variables is a really basic feature in most languages. And you actually disabled this important in order to simplify some rare edge-cases?
Is there any way to get this check in VS 2017?

**Comment by gafter** [comment]
@Epaminaidos We didn't actually "disable" anything here. It is already the case that we do not warn when the initializing expression is a computed value, and that behavior naturally applies to interpolated strings as well. You can file a feature request issue as I suggested if you want us to add a warning for this, but it is too late for us to add features to VS 2017. That has already shipped.

**Comment by Epaminaidos** [comment]
As far as I can see, this is already tagged as feature-request, isn't it?
Does it make any sense to add another one?
IMO, there should be a warning for each of the following variables:
`int i = 0;
Object o = new Object();
List<int> l = new List<int>();
l.Add(1);`

About "disabling": As far as I read in other sources, this has once been a feature that has been disabled on request of several developers.
It was not my intention to imply that this has been disabled in VS 2017 :-)

**Comment by gafter** [comment]
A proposal to revise the unused variable warning rules is https://github.com/dotnet/roslyn/issues/15695
